### ISSUE TITLE
Triton fallback server failure in workflow 10805.31 step 3

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#37767]
(moving from https://github.com/cms-sw/cmssw/issues/37738#issuecomment-1114455507)

The workflow 10805.31 step 3 fails with
```
Starting python2 /data/cmsbld/jenkins/workspace/ib-run-relvals/cms-bot/monitor_workflow.py timeout --signal SIGTERM 9000 cmsRun -j JobReport3.xml step3_RAW2DIGI_L1Reco_RECO_RECOSIM_PAT_VALIDATION_DQM.py
sh: nvidia-smi: command not found
%MSG-i ThreadStreamSetup: (NoModuleName) 02-May-2022 01:20:31 CEST pre-events
setting # threads 4
setting # streams 4
%MSG
02-May-2022 01:20:48 CEST Initiating request to open file file:step2.root
02-May-2022 01:20:50 CEST Successfully opened file file:step2.root
2022-05-02 01:21:27.078436: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations: SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
----- Begin Fatal Exception 02-May-2022 01:25:37 CEST-----------------------
An exception of category 'FallbackFailed' occurred while
[0] Calling beginJob
Exception Message:
Starting the fallback server failed with exit code 256
```
since CMSSW_12_4 2022-04-29-2300.

**Comment by makortel** [comment]
assign heterogeneous

**Comment by makortel** [comment]
@kpedro88 Can you take a look? Is `sh: nvidia-smi: command not found` a smoking gun?

**Comment by kpedro88** [comment]
No, that just means that GPUs aren't available and the fallback server will use the CPU. (It comes from https://github.com/cms-sw/cmssw/blob/master/HeterogeneousCore/SonicTriton/python/TritonService_cff.py#L11).

**Comment by kpedro88** [comment]
I only see this failure occurring in ppc/aarch IBs. In #35328, I disabled SonicTriton unit tests for non-amd64 OSes, because these require a different container that we do not yet have available.

**Comment by kpedro88** [comment]
Ah, those failures are the DRN unit test in 12_3_X (I had the wrong tab open in the IB dashboard). The linked comment does show the workflow failing in the IB matrix tests. @ssrothman can you take a look?

**Comment by Dr15Jones** [comment]
These RelVals are failing on x86

https://cmssdt.cern.ch/SDT/cgi-bin/logreader/el8_amd64_gcc10/CMSSW_12_4_X_2022-05-01-2300/pyRelValMatrixLogs/run/10805.31_SingleGammaPt35+2018_photonDRN+SingleGammaPt35_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt35+2018_photonDRN+SingleGammaPt35_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log#/

**Comment by ssrothman** [comment]
The workflow runs successfully on lpc interactive nodes (and iirc ran without problems in the PR tests). I'm not sure what could be causing it to fail in the IB. 

It looks like the failure is coming when spinning up the triton fallback server, which indicates to me that whatever is happening is happening in the sonic service rather than in the photonDRN producer. @kpedro88 what is exit code 256? Why might starting the fallback server fail?

**Comment by Dr15Jones** [comment]
Exit code 256 is a standard unit exit code for when the executable can not be found.

**Comment by ssrothman** [comment]
So maybe the IB unit test can not find the Triton executable/container with which to start up the fallback server?

**Comment by kpedro88** [comment]
Yes, I also tried running it on the lpc interactive nodes and was successful. @smuzaffar can you let one of us log into an IB test machine to try it there?

**Comment by smuzaffar** [comment]
I think error is random, for some IB it is `10805.31` and for other it is `10804.31` . Both if these workflows ran successfully for today's IB CMSSW_12_4 2022-05-02-1100

**Comment by smuzaffar** [comment]
@kpedro88 , try running on `cmsdev` nodes.

**Comment by kpedro88** [comment]
@Dr15Jones does your change in https://github.com/cms-sw/cmssw/pull/37764/files#diff-8899bb666d79547ed027cd8f543a4debcff63d07dde5211a87ca11527140adf3R269-R273 ensure that the output of the container launching script `cmsTriton` will be printed when an error occurs? That will probably make it easier to debug this in the future.

We could put this info directly into the exception message. I could also try to print the contents of the server logfile in the exception message.

**Comment by Dr15Jones** [comment]
I funnel the output of the execSys call to LogError so it will be printed. I didn't add it to the exception since I didn't know how big it would be.

**Comment by kpedro88** [comment]
The output of the command execution should only be a few lines. The server log will be a lot longer.

**Comment by kpedro88** [comment]
In the latest IB, the workflow failed for `slc7_amd64_gcc11` (succeeded in production arch `lc7_amd64_gcc10`).

Thanks to #37764, we get a little more information:
```
%MSG-e TritonService:  AfterModDestruction 03-May-2022 15:54:33 CEST pre-events
CMS_TRITON_GRPC_PORT: 8001
INFO:    instance started successfully
timed out waiting for server to start
INFO:    Stopping triton_server_instance_df61b571-1d28-480b-a54e-75df040f0d71 instance of /cvmfs/unpacked.cern.ch/.flat/87/874d87357f8f9b1b153d4d40a005349f5e2824d5cb46a54f24110f0f5d122c72 (PID=30808)
/cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_4_X_2022-05-03-1100/bin/slc7_amd64_gcc11/cmsTriton: line 275: 31499 Killed                  $DRYRUN singularity run instance://${SERVER} tritonserver $PORTARGS $REPOARGS $VERBOSE &>${REDIR}
```

It may be that loading the Singularity container over cvmfs is slow on some IB test nodes. Currently, the server start timeout is set to 120 seconds. Should we increase it to e.g. 300 seconds and see if that resolves the sporadic failures?

**Comment by makortel** [comment]
CMSSW_12_4_X_2022-05-03-1100 (that includes https://github.com/cms-sw/cmssw/pull/37764) slc7_amd64_gcc11 with
```
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
%MSG-e TritonService:  AfterModDestruction 03-May-2022 15:54:33 CEST pre-events
CMS_TRITON_GRPC_PORT: 8001
INFO:    instance started successfully
timed out waiting for server to start
INFO:    Stopping triton_server_instance_df61b571-1d28-480b-a54e-75df040f0d71 instance of /cvmfs/unpacked.cern.ch/.flat/87/874d87357f8f9b1b153d4d40a005349f5e2824d5cb46a54f24110f0f5d122c72 (PID=30808)
/cvmfs/cms-ib.cern.ch/nweek-02731/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_4_X_2022-05-03-1100/bin/slc7_amd64_gcc11/cmsTriton: line 275: 31499 Killed                  $DRYRUN singularity run instance://${SERVER} tritonserver $PORTARGS $REPOARGS $VERBOSE &>${REDIR}

%MSG
----- Begin Fatal Exception 03-May-2022 15:54:33 CEST-----------------------
An exception of category 'FallbackFailed' occurred while
   [0] Calling beginJob
Exception Message:
TritonService: Starting the fallback server failed with exit code 256
----- End Fatal Exception -------------------------------------------------
```
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc11/CMSSW_12_4_X_2022-05-03-1100/pyRelValMatrixLogs/run/10804.31_SingleGammaPt10+2018_photonDRN+SingleGammaPt10_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2018_photonDRN+SingleGammaPt10_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log#/

**Comment by qliphy** [comment]
CMSSW_12_4 2022-05-08-0000 [failure message](https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc10/CMSSW_12_4_X_2022-05-08-0000/pyRelValMatrixLogs/run/10805.31_SingleGammaPt35+2018_photonDRN+SingleGammaPt35_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt35+2018_photonDRN+SingleGammaPt35_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log#/35-35):

%MSG-e TritonService:  AfterModDestruction 08-May-2022 12:29:48 CEST pre-events
CMS_TRITON_GRPC_PORT: 8001
INFO:    instance started successfully
timed out waiting for server to start
INFO:    Stopping triton_server_instance_80dac23e-f01f-4c10-ae9c-689a8300f3e0 instance of /cvmfs/unpacked.cern.ch/.flat/87/874d87357f8f9b1b153d4d40a005349f5e2824d5cb46a54f24110f0f5d122c72 (PID=14943)
/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc10/cms/cmssw/CMSSW_12_4_X_2022-05-08-0000/bin/slc7_amd64_gcc10/cmsTriton: line 275: 15154 Killed                  $DRYRUN singularity run instance://${SERVER} tritonserver $PORTARGS $REPOARGS $VERBOSE &>${REDIR}

%MSG
----- Begin Fatal Exception 08-May-2022 12:29:48 CEST-----------------------
An exception of category 'FallbackFailed' occurred while
   [0] Calling beginJob
Exception Message:
TritonService: Starting the fallback server failed with exit code 256
----- End Fatal Exception -------------------------------------------------
08-May-2022 12:29:49 CEST  Closed file file:step2.root

**Comment by Dr15Jones** [comment]
Do not do 2. We already saw that the client side can generate long messages (which we are printing using LogError). If you REALLY want server logs they should also be handled via a LogError.

We do not want long exception messages as those are passed on to the workflow management system. In addition, long exception messages make it much harder to also just quickly understand what the problem is in general.

**Comment by Dr15Jones** [comment]
In addition, just increasing the timeout may reduce the error rate, but will not solve the underlying problem (which is still not fully understood).

**Comment by kpedro88** [comment]
I think printing the server logs is the only way to distinguish a Triton-specific problem from a cvmfs problem (too high load, slow downloading, etc.). I will submit a PR to put the server log into LogError.

**Comment by kpedro88** [comment]
see #37896

**Comment by kpedro88** [comment]
any feedback on increasing the wait time?

**Comment by smuzaffar** [comment]
@kpedro88 , no objections on increasing the timeout to 300s

**Comment by kpedro88** [comment]
see #38048

**Comment by makortel** [comment]
Copying @Dr15Jones' https://github.com/cms-sw/cmssw/pull/38048#issuecomment-1136062341

> This doesn't appear to have solved all the issues. We had a timeout failure in https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc10/CMSSW_12_5_X_2022-05-24-1100/pyRelValMatrixLogs/run/10804.31_SingleGammaPt10+2018_photonDRN+SingleGammaPt10_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano/step3_SingleGammaPt10+2018_photonDRN+SingleGammaPt10_pythia8_GenSimINPUT+Digi+RecoFakeHLT+HARVESTFakeHLT+ALCA+Nano.log#/187-187
> 
> The relevant part of the log appears to be
> 
> ```
> %MSG-w TritonFailure:  PatPhotonDRNCorrectionProducer:patPhotonsDRN  24-May-2022 13:44:45 CEST Run: 1 Event: 8
> An exception of category 'TritonFailure' occurred.
> Exception Message:
> evaluate(): unable to run and/or get result: Deadline Exceeded
> 
> %MSG
> 
> ----- Begin Fatal Exception 24-May-2022 13:44:45 CEST-----------------------
> An exception of category 'SonicCallFailed' occurred while
>    [0] Processing  Event run: 1 lumi: 1 event: 2 stream: 1
>    [1] Running path 'dqmofflineOnPAT_1_step'
>    [2] Prefetching for module SingleTopTChannelLeptonDQM_miniAOD/'singleTopElectronMediumDQM_miniAOD'
>    [3] Prefetching for module PATElectronSlimmer/'slimmedElectrons'
>    [4] Calling method for module PatPhotonDRNCorrectionProducer/'patPhotonsDRN'
> Exception Message:
> call failed after max 1 tries
> ----- End Fatal Exception -------------------------------------------------
> ```

**Comment by kpedro88** [comment]
This is a different error: the server started fine, but then the inference call did not complete in the maximum allotted time.

Right now, the timeout is set to a low value of 10 seconds: https://github.com/cms-sw/cmssw/blob/1bfd68cd1932e3351c7f120db16314c3d9867962/PhysicsTools/PatAlgos/python/slimming/patPhotonDRNCorrector_cfi.py#L12

@ssrothman can you suggest a better timeout value that will avoid spurious failures?

**Comment by makortel** [comment]
Seems that in the following IB, CMSSW_12_5_X_2022-05-24-2300, the workflow succeeded in all IB variants tested there, so at least the failure has become less frequent.

**Comment by smuzaffar** [comment]
@kpedro88 @ssrothman we still see this failure in 12.6.X IBs. Could it be that if multiple relvals run in parallel and try to start the triton inference server then only one will run and all other instances of triton server will fail?

**Comment by kpedro88** [comment]
@smuzaffar as I noted [above](https://github.com/cms-sw/cmssw/issues/37767#issuecomment-1136078310):
> This is a different error: the server started fine, but then the inference call did not complete in the maximum allotted time.

(The `cmsTriton` script tries to avoid conflicting parallel servers by looking for separate ports: https://github.com/cms-sw/cmssw/blob/38405a5b319be8ec094c981d6b45320aa577676a/HeterogeneousCore/SonicTriton/scripts/cmsTriton#L164-L182. This is not a perfect solution, but the problem here is different.)

**Comment by kpedro88** [comment]
@makortel this workflow has not failed in IBs in quite a while. [#49564](https://github.com/cms-sw/cmssw/pull/49564) may resolve the underlying problem, if my latest hypothesis was correct (that contention from pytorch automatically trying to use all available threads caused the inference to be unexpectedly slow on CPU). I propose closing this issue and opening a new one if the problem recurs.

**Comment by makortel** [comment]
+heterogeneous
